### PR TITLE
Fixed hang in XblCleanupAsync

### PR DESCRIPTION
**Pull request by mr-efficient**
CurlProvider::PerformAsync can be called simultaneously from multiple threads, which leaves CurlProvider::m_curlMultis in a bad state.  That results in m_curlMultis.size() being incorrect, which results in m_cleanupTasksRemaining being incorrect, which results in XAsyncComplete never being called in CurlProvider cleanup.

I protect against this by protecting access to m_curlMultis with m_mutex.

### CODE REVIEW

**Comment by mr-efficient** [comment]
One of the unit tests for our game hangs about 50% of the time on shutdown in XblCleanupAsync.  After a lot of breakpoint and printf debugging, I tracked the issue back to CurlProvider::m_curlMultis getting into a bad state:  The size returned by m_curlMultis.size() was different from the number of items returned by iterating from begin() to end().  I tracked _that_ back to CurlProvider::m_curlMultis being simultaneously accessed from multiple threads.

I'm not sure if protecting it with CurlProvider::m_mutex is the right fix, but it does seem to fix the hang.

File: Source/HTTP/Curl/CurlProvider.cpp [hunk 1]
```diff
@@ -81,13 +81,21 @@ HRESULT CurlProvider::PerformAsync(HCCallHandle hcCall, XAsyncBlock* async) noex
...
     RETURN_IF_FAILED(easyInitResult.hr);
 
-    auto iter = m_curlMultis.find(workPort);
-    if (iter == m_curlMultis.end())
-    {
-        auto multiInitResult = CurlMulti::Initialize(workPort);
-        RETURN_IF_FAILED(multiInitResult.hr);
+    http_internal_map<XTaskQueuePortHandle, HC_UNIQUE_PTR<xbox::httpclient::CurlMulti>>::iterator iter;
+
+	{
```
Comment (jplafonta) [inline Source/HTTP/Curl/CurlProvider.cpp:86]: >	 [](http://example.com/codeflow?start=0&length=1)

Please replace all tabs with spaces

File: Source/HTTP/Curl/CurlProvider.cpp [hunk 2]
```diff
@@ -133,28 +141,26 @@ HRESULT CALLBACK CurlProvider::CleanupAsyncProvider(XAsyncOp op, const XAsyncPro
...
-            {
-                // If there are no pending pending multi cleanups, complete cleanup here
-                cleanupComplete = true;
-            }
-        }
+		{
+			std::lock_guard<std::mutex> lock{ provider->m_mutex };
+			for (auto& pair : provider->m_curlMultis)
+			{
+				HRESULT hr = CurlMulti::CleanupAsync(std::move(pair.second), &provider->m_multiCleanupAsyncBlocks[multiIndex++]);
```
Comment (jplafonta) [inline Source/HTTP/Curl/CurlProvider.cpp:148]: >HRESULT hr = CurlMulti::CleanupAsync [](http://example.com/codeflow?start=4&length=36)

I believe it is not safe to call CurlMulti::CleanupAsync while holding m_mutex here.  CleanupAsync can complete synchronously in same cases, calling back into CurlProvider::MultiCleanupComplete which will try to grab the mutex.

CurlProvider::CleanupAsyncProvider has started, m_curlMultis should never be modified by another thread again - it should be safe to not lock here.

That said, to be safe we can grab the mutex, move m_curlMultis into a local, then release the lock and continue with cleanup as before but instead operating on the local curlMultis list

File: Source/HTTP/Curl/CurlProvider.cpp (hunk 1 above)
Comment (mr-efficient) [reply Source/HTTP/Curl/CurlProvider.cpp]: Done.

File: Source/HTTP/Curl/CurlProvider.cpp (hunk 2 above)
Comment (mr-efficient) [reply Source/HTTP/Curl/CurlProvider.cpp]: Okay, I've put up a revised version.  Thanks.

**Review by jplafonta — APPROVED** [review]
:shipit:

**Pull request merged**